core/capabilities/triggers/logevent/service.go

```diff
@@ -0,0 +1,161 @@
+var _ services.Service = &TriggerService{}
+
+// Creates a new Cron Trigger Service.
+// Scheduling will commence on calling .Start()
+func NewLogEventTriggerService(p Params) *TriggerService {
+	l := logger.Named(p.Logger, "LogEventTriggerCapabilityService")
+
+	logEventStore := NewCapabilitiesStore[logEventTrigger, capabilities.TriggerResponse]()
+
+	s := &TriggerService{
+		lggr:           l,
+		triggers:       logEventStore,
+		relayer:        p.Relayer,
+		logEventConfig: p.LogEventConfig,
+		stopChan:       make(services.StopChan),
+		wg:             sync.WaitGroup{},
+	}
+	s.CapabilityInfo, _ = s.Info(context.Background())
+	s.Validator = capabilities.NewValidator[RequestConfig, Input, capabilities.TriggerResponse](capabilities.ValidatorArgs{Info: s.CapabilityInfo})
+	return s
+}
+
+func (s *TriggerService) Info(ctx context.Context) (capabilities.CapabilityInfo, error) {
+	return capabilities.NewCapabilityInfo(
+		s.logEventConfig.Version(ID),
+		capabilities.CapabilityTypeTrigger,
+		"A trigger that listens for specific contract log events and starts a workflow run.",
+	)
+}
+
+// Register a new trigger
+// Can register triggers before the service is actively scheduling
+func (s *TriggerService) RegisterTrigger(ctx context.Context, req capabilities.TriggerRegistrationRequest) (<-chan capabilities.TriggerResponse, error) {
+	if req.Config == nil {
+		return nil, errors.New("config is required to register a log event trigger")
+	}
+	reqConfig, err := s.ValidateConfig(req.Config)
+	if err != nil {
+		return nil, err
+	}
+	// Add log event trigger with Contract details to CapabilitiesStore
+	respCh, err := s.triggers.InsertIfNotExists(req.TriggerID, func() (*logEventTrigger, chan capabilities.TriggerResponse, error) {
+		ctx, cancel := s.stopChan.NewCtx()
+		l, ch, err := newLogEventTrigger(ctx, cancel, s.lggr, req.Metadata.WorkflowID, reqConfig, s.logEventConfig, s.relayer)
+		s.wg.Add(1)
+		go func() {
+			defer s.wg.Done()
+			l.Listen(ctx) // Blocking call, until ctx.Done is issued
+		}()
+		return l, ch, err
+	})
+	if err != nil {
+		return nil, fmt.Errorf("LogEventTrigger %v", err)
+	}
+	s.lggr.Infow("RegisterTrigger", "triggerId", req.TriggerID, "WorkflowID", req.Metadata.WorkflowID)
+	return respCh, nil
+}
+
+func (s *TriggerService) UnregisterTrigger(ctx context.Context, req capabilities.TriggerRegistrationRequest) error {
+	trigger, ok := s.triggers.Read(req.TriggerID)
+	if !ok {
+		return fmt.Errorf("triggerId %s not found", req.TriggerID)
+	}
+	// Close callback channel and stop log event trigger listener
+	trigger.Stop()
+	// Remove from triggers context
+	s.triggers.Delete(req.TriggerID)
+	s.lggr.Infow("UnregisterTrigger", "triggerId", req.TriggerID, "WorkflowID", req.Metadata.WorkflowID)
+	return nil
+}
+
+// Start the service.
+func (s *TriggerService) Start(ctx context.Context) error {
+	if s.relayer == nil {
+		return errors.New("service has shutdown, it must be built again to restart")
```

Comment:
Is this error message accurate? Because it would be racey to mutate the `relayer` field after construction. It must not be set to nil.

Reply:
Removed this error now, since Start is not used, as of now